src/hotspot/share/jfr/leakprofiler/sampling/objectSampler.cpp

```diff
@@ -226,13 +226,13 @@ void ObjectSampler::sample(HeapWord* obj, size_t allocated, JavaThread* thread)
   }
   const JfrBlobHandle bh = get_thread_blob(thread, thread_id, virtual_thread);
   assert(bh.valid(), "invariant");
-  RecordStackTrace rst(thread);
   // try enter critical section
   JfrTryLock tryLock(&_lock);
   if (!tryLock.acquired()) {
     log_trace(jfr, oldobject, sampling)("Skipping old object sample due to lock contention");
     return;
   }
+  RecordStackTrace rst(thread);
```

Comment:
This change locks the sampler system for the duration the thread takes to generate its stack trace. During this time, no other thread can record with the sampler system, which is why the stack trace is taken outside of the lock.

What problem are you trying to solve? Stack traces are only tuned on in profile.jfc which is opt-in and comes with a higher overhead.